### PR TITLE
[package] cgns/4.3.0: Updated release, plus some recipe fixes

### PR DESCRIPTION
**Pull request by paulharris**
Specify library name and version:  **cgns/4.3.0**

Required for future VTK package, as an external dependency.

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by uilianries** [comment]
Probably that new release is not prepared to work with clang13. It would need a patch, but accepted by the upstream too. Also, another possibility is skipping that broken configuration.

**Comment by paulharris** [comment]
Ok.  I've continued to work on this, trying to modernise it.  Having some issues.

**Comment by paulharris** [comment]
I imagine this will build when #10836 is merged.
How do I trigger a CI retry when that time comes?  Is it somehow automatic?

**Comment by uilianries** [comment]
>  How do I trigger a CI retry when that time comes? Is it somehow automatic?

No. You need to do it manually.

Usually we mark with `blocked` label and add the comment `depends on #<issue_number>`

**Comment by paulharris** [comment]
depends on #10836

**Comment by paulharris** [comment]
@SSE4 or anyone else, can someone please take a look at why this isn't building in the CI ?
It should build without problems...

**Comment by SSE4** [comment]
@paulharris I've checked, and locally I have exactly the same error as CI:
```
CMake Error at source_subfolder/src/CMakeLists.txt:78 (string):
string sub-command COMPARE, mode EQUAL needs at least 5 arguments total to
command.                                                                         
```

**Comment by paulharris** [comment]
> @paulharris I've checked, and locally I have exactly the same error as CI:
> 
> ```
> CMake Error at source_subfolder/src/CMakeLists.txt:78 (string):
> string sub-command COMPARE, mode EQUAL needs at least 5 arguments total to
> command.                                                                         
> ```

Thanks @SSE4 !

**Comment by paulharris** [comment]
I undid my major conan-v2 changes and have only changed what the Linter asked for.

**Comment by paulharris** [comment]
(snipped)
>       small_orange_diamond Informative: This recipe is not ready for Conan v2
>       ```
>       ImportError: cannot import name 'CMake' from 'conans' (/opt/pyenv/versions/3.7.13/lib/python3.7/site-packages/conans/__init__.py)
>       ```

I cannot fix this warning with conan 1.50.0,
changing the import to `from conan.tools.cmake import CMake`
Results in this error message:
```
ERROR: cgns/4.3.0@cci/test: Error in build() method, line 107
        cmake = self._configure_cmake()
while calling '_configure_cmake', line 80
        self._cmake = CMake(self)
        ConanException: Usage of toolchain is only supported with 'cmake_find_package' or 'cmake_find_package_multi' generators
```

File: recipes/cgns/all/conanfile.py [hunk 1]
```diff
@@ -37,23 +40,37 @@ def _source_subfolder(self):
...
+
     def configure(self):
         if self.options.shared:
             del self.options.fPIC
         del self.settings.compiler.libcxx
         del self.settings.compiler.cppstd
 
+        if self.options.parallel:
+            self.options["hdf5"].parallel = True
+            self.options["hdf5"].enable_cxx = False # can't enable this with parallel
```
Comment (prince-chrismc) [inline recipes/cgns/all/conanfile.py:64]: this would need to be in the validate as well

Comment (paulharris) [reply recipes/cgns/all/conanfile.py:64]: As well, or instead-of ?
Should this recipe be setting hdf5's options like this?
Or should we simply validate to check the consumers have set the appropriate options in HDF5 ?

File: recipes/cgns/all/conanfile.py [hunk 2]
```diff
@@ -85,6 +113,7 @@ def package(self):
         cmake.install()
 
         os.remove(os.path.join(self.package_folder, "include", "cgnsBuild.defs"))
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
```
[the remark this answers is not in the document]
Comment (paulharris) [reply recipes/cgns/all/conanfile.py:116]: Hrm, so as the recipe stands, it will generate the target cgns::cgns, right?

Here is a repo with all the outputs from a shared+static build, could you please help to look at the targets generated?
https://github.com/paulharris/test-cgns-output

From what I can tell, it also generates:
* CGNS::cgns_static (for static builds)
* CGNS::cgns_shared (for shared builds)

`foreach(_expectedTarget CGNS::cgns_static)`

For shared, it also adds a define.  I've dealt with that in a different way in package_info
```
set_target_properties(CGNS::cgns_shared PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "USE_DLL"
```

So I should add components for these?
I'd like to stay with cgns::cgns for the "normal" target...
I'm not sure I should provide these other targets as they are the "old way".

**Comment by paulharris** [comment]
@prince-chrismc gentle ping - just waiting on follow-ups from you when you have time

File: recipes/cgns/all/conanfile.py (hunk 1 above)
Comment (SSE4) [reply recipes/cgns/all/conanfile.py:64]: in general, it's not recommended to modify optiions like that
instead, invalid combinations are rejected in `validate`
but sometimes it's not possible (e.g. because of necessity to use `shared` dep while `static` is default one)

File: recipes/cgns/all/conanfile.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (SSE4) [reply recipes/cgns/all/conanfile.py:116]: if you aim for flawless (=transparent) integration, you better provide these targets
for instance, if some library depends on `cgns` and uses `find_package(CGNS)` + `target_link_libraries(${PROJECT_NAME} PRIVATE CGNS::cgns_static)` it won't work out of the box with conan generator - you would need to patch `CMakeLists.txt`.
it's not mandatory to provide these components, but in general it's good practice. however, you can add it later, on demand, in another PR, it doesn't block an acceptance to CCI. (it's better to have recipe with no components over having no recipe at all).

**Review by SSE4 — DISMISSED** [review]

**Comment by paulharris** [comment]
I don't know how to migrate the imports to conan.tools.files.CMake
locally, it complains:
```
ERROR: cgns/4.3.0@ccitest/run-a: Error in build() method, line 101
        cmake = self._configure_cmake()
        while calling '_configure_cmake', line 78
                self._cmake = CMake(self)
                        ConanException: Usage of toolchain is only supported with 'cmake_find_package' or 'cmake_find_package_multi' generators
```

**Comment by paulharris** [comment]
This now uses conan v2 techniques, as best as I can tell,
and builds locally with conan 1.50.0

But, I don't know I have done it right, and I'm not sure how it can be done better.
Please give me some feedback !

File: recipes/cgns/all/conanfile.py [hunk 3]
```diff
@@ -31,80 +30,71 @@ class CgnsConan(ConanFile):
...
+        tc.variables["CGNS_ENABLE_FORTRAN"] = False
+        tc.variables["CGNS_ENABLE_HDF5"] = self.options.with_hdf5
+        tc.variables["CGNS_BUILD_SHARED"] = self.options.shared
+        tc.variables["CGNS_USE_SHARED"] = self.options.shared
+        tc.variables["CGNS_ENABLE_PARALLEL"] = self.options.parallel
+        tc.variables["CGNS_BUILD_CGNSTOOLS"] = False
         # CMAKE_BUILD_TYPE is not set by all the CMake generators, but cgns needs it
-        self._cmake.definitions["CMAKE_BUILD_TYPE"] = self.settings.build_type
-        self._cmake.configure()
+        tc.variables["CMAKE_BUILD_TYPE"] = self.settings.build_type
```
Comment (prince-chrismc) [inline recipes/cgns/all/conanfile.py:79]: I'd expect conan to handle this 🤔

Comment (paulharris) [reply recipes/cgns/all/conanfile.py:79]: Ok great, I'll chop out that line.

File: recipes/cgns/all/conanfile.py (hunk 1 above)
Comment (paulharris) [reply recipes/cgns/all/conanfile.py]: I assume the method I use now is ok?

File: recipes/cgns/all/conanfile.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (paulharris) [reply recipes/cgns/all/conanfile.py]: Ok, should be ok now?

**Comment by paulharris** [comment]
Is this current error because I don't list libdirs in package() ?

**Comment by paulharris** [comment]
I'm baffled... the log file says:

```
-- Installing: /home/conan/w/prod/BuildSingleReference/.conan/data/cgns/4.3.0/_/_/package/e23c43bec4aabdcbdb4d445bceb4d028b4e15822/lib/libcgns.a
....
[HOOK - conan-center.py] post_package_info(): ERROR: [LIBRARY DOES NOT EXIST (KB-H054)] Component cgns::cgns_static library 'cgns' is listed in the recipe, but not found installed at self.cpp_info.libdirs. Make sure you compiled the library correctly. If so, then the library name should probably be fixed. Otherwise, then the component should be removed. (https://github.com/conan-io/conan-center-index/blob/master/docs/error_knowledge_base.md#KB-H054) 
```

I added libdirs to the components just to try ...

File: recipes/cgns/all/test_package/CMakeLists.txt [hunk 4]
```diff
@@ -1,8 +1,7 @@
 cmake_minimum_required(VERSION 3.1)
 project(test_package CXX)
 
-include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
-conan_basic_setup()
+find_package(CGNS REQUIRED)
 
 add_executable(test_package test_package.cpp)
-target_link_libraries(test_package ${CONAN_LIBS})
+target_link_libraries(test_package CGNS::CGNS)
```
Comment (prince-chrismc) [inline recipes/cgns/all/test_package/CMakeLists.txt:7]: Because there are components I assume the headers are now being passed to the global target 🤔

Comment (paulharris) [reply recipes/cgns/all/test_package/CMakeLists.txt:7]: I don't follow.
Is CGNS::CGNS the "global target" ?
What should be changed?

Comment (paulharris) [reply recipes/cgns/all/test_package/CMakeLists.txt:7]: @prince-chrismc when you have time please, I'm not sure what you mean

Comment (prince-chrismc) [reply recipes/cgns/all/test_package/CMakeLists.txt:7]: Sorry for not being more clear, I was trying to see why the build fail when I left that comment...

CI failed because the header was not found, but I can see it being placed in the package... I am not sure why that is I was just tossing a suggestion... might be a bug that was fix in 1.50 thats not in CCI

🤷 sorry I am not very helpful 🤗 💟

Comment (paulharris) [reply recipes/cgns/all/test_package/CMakeLists.txt:7]: Ok thanks mate.
@SSE4 could you suggest someone who could help with this?  (I'm not sure who I should tag as Triage Person)

File: recipes/cgns/all/conandata.yml [hunk 5]
```diff
@@ -1,10 +1,8 @@
 sources:
-  "3.4.1":
```
Comment (uilianries) [inline recipes/cgns/all/conandata.yml:2]: Could you please explain why did you need to drop 3.4.1?

File: recipes/cgns/all/conanfile.py [hunk 6]
```diff
@@ -1,92 +1,125 @@
...
-
-    @property
-    def _source_subfolder(self):
-        return "source_subfolder"
-
-    @property
-    def _build_subfolder(self):
-        return "build_subfolder"
+    def export_sources(self):
+        copy(self, "CMakeLists.txt", src=".", dst=self.export_sources_folder)
```
Comment (uilianries) [inline recipes/cgns/all/conanfile.py:33]:
```suggestion
```
No more cmake wrapper! freedom!

File: recipes/cgns/all/conanfile.py [hunk 7]
```diff
@@ -1,92 +1,125 @@
...
+            raise ConanInvalidConfiguration("The option 'parallel' requires HDF5 with parallel=True")
+        if self.info.options.parallel and self.info.options.with_hdf5 and self.info.options["hdf5"].enable_cxx:
+            raise ConanInvalidConfiguration("The option 'parallel' requires HDF5 with enable_cxx=False")
+
     def configure(self):
         if self.options.shared:
             del self.options.fPIC
-        del self.settings.compiler.libcxx
+        try:
+            # In windows, with msvc, the compiler.libcxx doesn't exist, so it will raise.
```
Comment (uilianries) [inline recipes/cgns/all/conanfile.py:51]: which exception does it raise? I'm curious because Visual Studio does not have libcxx too and so far, we don't have such problem. Maybe, it's a case to be compatible.

Comment (prince-chrismc) [reply recipes/cgns/all/conanfile.py:51]: I think its a 2.0 change where it now fails when the settings are missing (I don't recall where I read that though)

File: recipes/cgns/all/conandata.yml (hunk 5 above)
Comment (paulharris) [reply recipes/cgns/all/conandata.yml:2]: I couldn't make it work with conan v1-v2 transition, I'm close to giving up ...
I am happy to not drop 3.4.1 but I can't keep spending time bashing my head against walls.

It is probably something very simple for someone else to fix...

**Comment by paulharris** [comment]
I'll continue this once I get the other packages working.

File: recipes/cgns/all/conanfile.py (hunk 6 above)
Comment (paulharris) [reply recipes/cgns/all/conanfile.py:33]: yay!

**Comment by paulharris** [comment]
Ok, and we are back.  This may need the HDF5 package to be merged first, which may require szip ... there are a few pending PRs.

**Comment by paulharris** [comment]
I don't understand why mac for cgns:shared=True fails, and it works everywhere else:
https://github.com/eirikb/proof-of-conan/actions/runs/3298518790/jobs/5440745006

**Review by prince-chrismc — APPROVED** [review]

**Comment by prince-chrismc** [comment]
You are getting good at these paul ❤️ welcome back 😉

**Review by jwillikers — APPROVED** [review]
Thanks for working so hard on this!

**Comment by paulharris** [comment]
Thanks for the encouragement guys!
